src/ophyd_async/core/_device.py

```diff
@@ -47,10 +71,12 @@ def log(self):
             getLogger("ophyd_async.devices"), {"ophyd_async_device_name": self.name}
         )
 
-    def children(self) -> Iterator[tuple[str, "Device"]]:
-        for attr_name, attr in self.__dict__.items():
-            if attr_name != "parent" and isinstance(attr, Device):
-                yield attr_name, attr
+    def children(self) -> dict[str, Device]:
+        return {
+            attr_name: attr
+            for attr_name, attr in self.__dict__.items()
+            if attr_name != "parent" and isinstance(attr, Device)
+        }
```

Comment:
Tbh we don't have to define use an abstract class, instead we can:

``` python
class Device(HasName):
    @property
    def children(self) -> "DeviceVector":
        return DeviceVector(
            {
                attr_name: attr
                for attr_name, attr in self.__dict__.items()
                if attr_name != "parent" and isinstance(attr, Device)
            }
        )
    def connect(self, ...):
        # DO STUFF
        self.children.connect(...)

class DeviceVector(dict[int, VT], Device):
    """
    Defines device components with indices.

    In the below example, foos becomes a dictionary on the parent device
    at runtime, so parent.foos[2] returns a FooDevice. For example usage see
    :class:`~ophyd_async.epics.demo.DynamicSensorGroup`
    """

    def connect(self, *args, **kwargs):
        # Logic for connecting which can be used for by Device
        # For its children
        gather({
            device.connect(*args, **kwargs)
            for device_name, device in self.items():
        })
        

    def children(self) -> DeviceVector:
        return self
```

Reply:
This reduces the code a bit for the current use case, but I think it will break what I had planned for PVI. Let me write that first then we can revisit this and see if this still applies...